### PR TITLE
Add YAML files for land DA

### PR DESCRIPTION
**Pull request by jiaruidong2017**
https://github.com/NOAA-EMC/GDASApp/issues/136
Add YAML files (LETKF-OI and OBS LIST) to the GDASApp for land DA.

### CODE REVIEW

File: parm/land/hofx/hofx_nomodel.yaml
Comment (CoryMartin-NOAA) [inline parm/land/hofx/hofx_nomodel.yaml]: We probably also need this and the following to be `$(LAND_BKG_ISOTIME)` and `$(LAND_BKG_YYYYmmddHHMMSS)`? Just for the added flexibility?

File: parm/land/letkfoi/letkfoi.yaml
Comment (CoryMartin-NOAA) [inline parm/land/letkfoi/letkfoi.yaml]: Same comment on the time variables

File: parm/land/letkfoi/letkfoi.yaml
Comment (CoryMartin-NOAA) [inline parm/land/letkfoi/letkfoi.yaml]: Didn't @ClaraDraper-NOAA say to uses the templating here?

**Review by CoryMartin-NOAA — COMMENTED** [review]
Few small comments

File: parm/land/letkfoi/letkfoi.yaml
Comment (ClaraDraper-NOAA) [inline parm/land/letkfoi/letkfoi.yaml:1]: I think we need a specific "field metadata override" within the geometry block for the land DA.

**Review by ClaraDraper-NOAA — COMMENTED** [review]
Thanks Jiarui - before finalizing this, can you please create a yaml file for the test case in the land-offline_workflow, and check that is passes? You can override my yaml constructor by specifying the yaml name.

**Comment by ClaraDraper-NOAA** [comment]
@jiaruidong2017 Note also that I'm currently doing hofx using the letkf executable, but here it looks like you're using the nomodel executable. I switched to using the letkf executable so that I didn't have to maintain two yaml structures. I suppose it's OK to switch back if we're going to use the GDASApp to generate the yaml.

File: parm/land/hofx/hofx_nomodel.yaml
Comment (jiaruidong2017) [reply parm/land/hofx/hofx_nomodel.yaml]: Updated

File: parm/land/letkfoi/letkfoi.yaml
Comment (jiaruidong2017) [reply parm/land/letkfoi/letkfoi.yaml]: Updated

File: parm/land/letkfoi/letkfoi.yaml
Comment (jiaruidong2017) [reply parm/land/letkfoi/letkfoi.yaml]: Made changes to use `template` format.

File: parm/land/letkfoi/letkfoi.yaml
Comment (jiaruidong2017) [reply parm/land/letkfoi/letkfoi.yaml:1]: @ClaraDraper-NOAA This `field metadata override` is included in the geom dictionary `GEOM_BKG`. We can add the entry `gfs-land.yaml` for our land DA in the yamltool later.

**Comment by jiaruidong2017** [comment]
> @jiaruidong2017 Note also that I'm currently doing hofx using the letkf executable, but here it looks like you're using the nomodel executable. I switched to using the letkf executable so that I didn't have to maintain two yaml structures. I suppose it's OK to switch back if we're going to use the GDASApp to generate the yaml.

@ClaraDraper-NOAA This hofx using nomodel executable is a place holder, because all other components have this doing hofx for nomodel. We can add the hofx using letkf executable later. 
@CoryMartin-NOAA Please correct me if it is not the case.

**Comment by jiaruidong2017** [comment]
> Thanks Jiarui - before finalizing this, can you please create a yaml file for the test case in the land-offline_workflow, and check that is passes? You can override my yaml constructor by specifying the yaml name.

@ClaraDraper-NOAA Thanks for your comments. I can do the test in the land-offline_workflow when I updated the yamltool to generate a final yaml file from GDASApp.

**Comment by ClaraDraper-NOAA** [comment]
@jiaruidong2017 It looks like everything here has been addressed (right?). Once you can confirm that the yaml generated by the GDASApp can be used in in my workflow, I'll approve the merge. 

@CoryMartin-NOAA At the moment I am maintaining builds of fv3-bundle and ioda-converters. However, my build may get out of synch with the one GDASApp is using, so I will switch to using the same versions as you. Are GDASApp users all pointing to the same build, or maintaining their own versions?

**Comment by CoryMartin-NOAA** [comment]
@ClaraDraper-NOAA PR #146 will go in today/Tuesday. It switches to using tags/commits for JEDI repos instead of `develop`. I think that will hopefully make it easy to ensure the builds are the same.

**Comment by ClaraDraper-NOAA** [comment]
> @ClaraDraper-NOAA PR #146 will go in today/Tuesday. It switches to using tags/commits for JEDI repos instead of `develop`. I think that will hopefully make it easy to ensure the builds are the same.

Fantastic, this will be much easier. I'll switch to using the same version. Will you have an authoritative installation on hera that I can point to? Or should I install my own version for the land DA workflow to point to?

**Comment by CoryMartin-NOAA** [comment]
@ClaraDraper-NOAA no authoritative install planned at this point (but perhaps eventually?), so for now I would install your own version. That PR includes switching to `spack-stack` as well

**Comment by ClaraDraper-NOAA** [comment]
Got it, thanks.

**Comment by jiaruidong2017** [comment]
@ClaraDraper-NOAA Yes, I believe I have addressed all the questions. I am currently working on the yamltool to generate the final yaml file for the land DA. However, it will take some time.
In addition, all the yaml files used in the GDASApp came from the `land-offline_workflow`, which already passed both ims snow DA and gts snow DA tests in the `land-offline_workflow`.

**Comment by ClaraDraper-NOAA** [comment]
@jiaruidong2017 To approve this, I'm waiting for you to use the GDASApp to create a land DA yaml, and then confirm that that yaml is correct (e.g., can be used in the offline system).

**Comment by jiaruidong2017** [comment]
@CoryMartin-NOAA The current `run_jedi_exe` in the GDASApp just supported  three modes (`hofx`, `variational`, `gw_scripts`). For the land DA, we currently use the `letkf` mode. To get the final yaml file for the land DA, the `letfk` mode needs to be added to the supported modes in the GDASApp. Is it correct? I would appreciate your any comments and suggestions. Thank you very much.

**Comment by CoryMartin-NOAA** [comment]
@jiaruidong2017 the `run_jedi_exe` script is to run a stand-alone analysis for testing and is not explicitly required at this time to include this for your work. The YAML should be able to be produced by the `ush/genYAML` script and should match what goes into FV3-JEDI (at least that is what I think @ClaraDraper-NOAA is asking you to do)

**Comment by jiaruidong2017** [comment]
> @jiaruidong2017 To approve this, I'm waiting for you to use the GDASApp to create a land DA yaml, and then confirm that that yaml is correct (e.g., can be used in the offline system).

@CoryMartin-NOAA @ClaraDraper-NOAA I have generated the yaml file, and conducted the snow DA tests successfully. I confirm the derived yaml file is correct, and I pushed the changes into the GDASApp repo for your review. Please let me know if you have any questions. Thanks.

File: parm/land/obs/config/adpsfc_snow.yaml
Comment (CoryMartin-NOAA) [inline parm/land/obs/config/adpsfc_snow.yaml]: I think we probably want to follow a convention more akin to what is in ops now, suggest:
`$(OBS_DIR)/$(OBS_PREFIX)adpsfc_snow.$(OBS_DATE).nc4`

So that it can be `gdas` and `gfs` if we want.

File: parm/land/obs/lists/gdas_land_prototype.yaml
Comment (CoryMartin-NOAA) [inline parm/land/obs/lists/gdas_land_prototype.yaml:2]: Just confirming, this is the only ob that will be used (at least for initial prototyping)?

File: ush/ufsda/yamltools.py
Comment (CoryMartin-NOAA) [inline ush/ufsda/yamltools.py]: one is () and the other is {}, I don't care which for this error print but make it consistent

File: ush/ufsda/yamltools.py
Comment (CoryMartin-NOAA) [inline ush/ufsda/yamltools.py]: is this any different than the atm case? Is there a way we can reuse one?

File: ush/ufsda/yamltools.py
Comment (CoryMartin-NOAA) [inline ush/ufsda/yamltools.py]: We will need to stage these in a fix dir before merging it in, we will want to support both hera and orion

File: ush/ufsda/yamltools.py
Comment (CoryMartin-NOAA) [inline ush/ufsda/yamltools.py]: This needs to be resolution dependent, right?

**Review by CoryMartin-NOAA — COMMENTED** [review]
A few more comments, but it's mostly coming together I think

File: parm/land/obs/config/adpsfc_snow.yaml
Comment (jiaruidong2017) [reply parm/land/obs/config/adpsfc_snow.yaml]: Added $(OBS_PREFIX). Thanks.

File: parm/land/obs/lists/gdas_land_prototype.yaml
Comment (jiaruidong2017) [reply parm/land/obs/lists/gdas_land_prototype.yaml:2]: Added the second one (ims snow).

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py]: Made change for keeping the consistency.

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py]: Yes, I made the changes.

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py]: The only difference is at the end section to call `fv3land_geom_dict` as below: 

```
config['GEOM_BKG'] = fv3land_geom_dict(case, levs, ntiles, layout, io_layout)
config['GEOM_ANL'] = fv3land_geom_dict(case_anl, levs, ntiles, layout, io_layout)
```
There are many differences in the `fv3land_geom_dict` compared with `fv3_geom_dict`. The reason I added the `landanl_case` is to show the difference in name for land analysis from the `atmanl_case` for atm analysis. If you want me to reuse the `atmanl_case` for the land, I will make the corresponding changes. Thanks.

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py]: Is it okay for the fix dir like `/scratch2/NCEPDEV/land/data/fix/` on hera?

File: parm/land/letkfoi/letkfoi.yaml
Comment (jiaruidong2017) [reply parm/land/letkfoi/letkfoi.yaml:1]: `field metadata override: gfs-land.yaml` is added to the yamltools.py for the land DA.

File: ush/ufsda/yamltools.py
Comment (CoryMartin-NOAA) [reply ush/ufsda/yamltools.py]: Yeah I think we rename it to 'fv3anl_case' or something like that and just add the differing `if` statement, that way we aren't duplicating 80% of the routine again. I think this change will then cover atm, land, and aero with only minor additions needed for their specific tasks.

File: ush/ufsda/yamltools.py
Comment (CoryMartin-NOAA) [reply ush/ufsda/yamltools.py]: No, eventually we will want all the fix files with the other fix files, and these are maintained by the global-workflow team. This is fine for testing, but making a note here that this will have to be changed before we add land DA to the workflow itself.

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py]: As suggested to reduce the duplications, I renamed the `atmanl_case` to `fv3anl_case`, removed the `landanl_case`, and added the `landanl_case` into the `fv3anl_case`.

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py]: Add the comment line making a note to indicate future changes required before we add land DA to the workflow.

File: parm/land/letkfoi/letkfoi.yaml
Comment (ClaraDraper-NOAA) [inline parm/land/letkfoi/letkfoi.yaml]: The number of members should be hard-coded to 2 for the LETKFOI.

File: parm/land/obs/config/adpsfc_snow.yaml
Comment (ClaraDraper-NOAA) [inline parm/land/obs/config/adpsfc_snow.yaml:19]: Out of curiosity: I haven't see this before. Why are there the two components? What does specifying the BackgroundErrorIdentity here do?

File: ush/ufsda/yamltools.py
Comment (ClaraDraper-NOAA) [reply ush/ufsda/yamltools.py]: This path was just intended as a temporary one for my tests while EMC finalized the fractional grids. I recommend changing this to the fix directory with the orography files that are being used by the rest of the work flow, since the land DA should be using the same files as the atmos DA.

File: ush/ufsda/yamltools.py
Comment (ClaraDraper-NOAA) [inline ush/ufsda/yamltools.py:169]: What is the ISOTIME?

File: ush/ufsda/yamltools.py
Comment (ClaraDraper-NOAA) [inline ush/ufsda/yamltools.py:189]: 3 questions: 
1. Should BEGIN_YYYYmmddHHMMSS have an ATM_ prefix, for consistency with the variables above? 
2. Is there any difference between the land and atm variables. If not, why do we need separate land variables. 
3. The land DA is applied to a restart - for DA with IAU, the restarts are only available at the start of the DA window (and not the middle). I'm not sure how the valid_time is defined - will this give us the increment at a time that we have a restart?

File: parm/land/obs/config/adpsfc_snow.yaml
Comment (ClaraDraper-NOAA) [reply parm/land/obs/config/adpsfc_snow.yaml]: @CoryMartin-NOAA When would something be labeled GFS and when would it be GDAS? Are there any other options?

File: ush/ufsda/yamltools.py
Comment (ClaraDraper-NOAA) [reply ush/ufsda/yamltools.py]: Also, do we need to change the orog file name to include the ocean details (mxXX) -  or again, just copy what's done on the atmos side.

**Review by ClaraDraper-NOAA — CHANGES_REQUESTED** [review]
Thanks @jiaruidong2017 . I have a couple of small code change requests, and then some questions to help me understand how the GDAS app works.

File: ush/ufsda/yamltools.py
Comment (CoryMartin-NOAA) [reply ush/ufsda/yamltools.py:169]: This distinguishes in the YAML `20221031.120000` (like for FV3) and the ISO8601 string `2022-10-31T12:00:00Z`

File: ush/ufsda/yamltools.py
Comment (CoryMartin-NOAA) [reply ush/ufsda/yamltools.py:189]: 1) probably good to be safe
2) I guess the question is, would there ever be a land analysis at a different resolution from atmosphere? Probably not?
3) This is something we still need to sort out, for this initial go, it will be 3DVar everywhere. When I have a solution for IAU, it will be updated appropriately.

File: parm/land/obs/config/adpsfc_snow.yaml
Comment (CoryMartin-NOAA) [reply parm/land/obs/config/adpsfc_snow.yaml]: GFS = early cycle 
GDAS = late cycle

That should be the only options for global, it helps distinguish which obs are available (latency, etc.)

File: ush/ufsda/yamltools.py
Comment (CoryMartin-NOAA) [reply ush/ufsda/yamltools.py]: I think it's just the atm resolution for the orography but I don't know the details about the fractional grid files. I would just be consistent with what the coupled model is using on the atmos side if possible.

Comment (ClaraDraper-NOAA) [reply ush/ufsda/yamltools.py]: The fractional grids have slightly different orography, and different files names. I think the main thing here is just to make sure that this is the same as for the atmos DA / and model.

File: ush/ufsda/yamltools.py
Comment (ClaraDraper-NOAA) [reply ush/ufsda/yamltools.py:189]: I think(?) that for IAU the timing of the restarts is different, so we may eventually need some options in here to change the timing of the land DA update within the window. If there's a restart available at valid_time, then we're good for now.

File: parm/land/letkfoi/letkfoi.yaml
Comment (jiaruidong2017) [reply parm/land/letkfoi/letkfoi.yaml]: The number of members is assigned in the `genYAML_prep.sh` together with all other assigned variables, such as time, resolution etc.

File: parm/land/obs/config/adpsfc_snow.yaml
Comment (jiaruidong2017) [reply parm/land/obs/config/adpsfc_snow.yaml:19]: This is required by the Buddy Check filter.

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py:169]: Thanks Cory.

File: parm/land/letkfoi/letkfoi.yaml
Comment (jiaruidong2017) [reply parm/land/letkfoi/letkfoi.yaml]: @ClaraDraper-NOAA As you requested during today's meeting, I made the change to hard-code to 2 for the LETKFOI.

File: ush/ufsda/yamltools.py
Comment (ClaraDraper-NOAA) [reply ush/ufsda/yamltools.py]: @jiaruidong2017 Can you please change this to use the same fix files as the atmos DA?

File: ush/ufsda/yamltools.py
Comment (ClaraDraper-NOAA) [reply ush/ufsda/yamltools.py:189]: @jiaruidong2017 - can you response to this?

**Comment by ClaraDraper-NOAA** [comment]
@jiaruidong2017 You still need to change the fix directory, and address the questions about the timing of the window.

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py]: @CoryMartin-NOAA Do you know where I can find the fix files for the atmos DA on hera? I  can do a check for the orography. Thanks.

Comment (CoryMartin-NOAA) [reply ush/ufsda/yamltools.py]: we don't use fix files (as of now) for orography for DA, but can we point to the model fix files that Kate Friedman et al. maintain? I assume that is what you want (what FV3 uses for orography). Or are the fields you need in the model RESTARTs?

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py:189]: @ClaraDraper-NOAA I think the IAU timing of the restarts for the GTS snow DA is consistent with the atm DA, which is in the middle of the DA window. For the GTS snow DA, we use the snow depth observations from the adpsfc dump files, which are same as the atm DA to use the conventional dump or prepbufr files. Please let me know if you have different idea.

File: ush/ufsda/yamltools.py
Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py]: Thanks @CoryMartin-NOAA 
@barlage @KateFriedman-NOAA Do you know where we can find the FV3 model fix files for the variable `orography` maintained by the global workflow on hera? Thanks.

Comment (KateFriedman-NOAA) [reply ush/ufsda/yamltools.py]: @jiaruidong2017 The fix file set on Hera is here:

/scratch1/NCEPDEV/global/glopara/fix

The orography subset is:

/scratch1/NCEPDEV/global/glopara/fix/orog/20220805/

Note, the `20220805` subfolder is the current version for the orog files. When updates are made to those files a new timestamp (YYYYMMDD) subfolder will be created and users will be directed to point there for the orog files.

FYI, the global-workflow handles the fix file symlinks within GFS clones using a version file that a link script sources when setting up the clone for running. See these if curious:

https://github.com/NOAA-EMC/global-workflow/blob/develop/sorc/link_workflow.sh
https://github.com/NOAA-EMC/global-workflow/blob/develop/versions/fix.ver

Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py]: @KateFriedman-NOAA Thank you very much for your information.

**Comment by jiaruidong2017** [comment]
> @jiaruidong2017 You still need to change the fix directory, and address the questions about the timing of the window.

@ClaraDraper-NOAA @CoryMartin-NOAA I have updated the fix directory to point out to the `glopara` fix directory for orography, and the snow DA test was successful with new generated yaml file. The questions regarding the timing of the window was addressed above. Please let me know if you have any questions. Thanks.

File: ush/ufsda/yamltools.py
Comment (ClaraDraper-NOAA) [reply ush/ufsda/yamltools.py:189]: @jiaruidong2017 Have you added the ATM prefix to BEGIN_YYYYmmddHHMMSS? See point 1 above. 

For the timing, the (possible) issue is that the land DA update is applied to a restart. So we should make the analysis time be the same as the restart time - which means it may not be the same as the atmos analysis time. For example,  for IAU the restart is at the start of the window and I think (?) the (nominal) atmos analysis time is the middle of the window. In this case the land analysis time should be the start of the window (or, the end of the previous window).

Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py:189]: > Have you added the ATM prefix to BEGIN_YYYYmmddHHMMSS? See point 1 above.

@ClaraDraper-NOAA I couldn't make this change, because this is belong to the atm DA. If making this change, the yaml files in the `parm/atm/` will require the corresponding changes. I don't have authority to make any changes for the atm DA. @CoryMartin-NOAA Am I correct?

Comment (CoryMartin-NOAA) [reply ush/ufsda/yamltools.py:189]: @jiaruidong2017 no feel free to make the change, as long as it is changed everywhere it shouldn't break anything. @ClaraDraper-NOAA is right about the logic here on why it is needed.

Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py:189]: @ClaraDraper-NOAA @CoryMartin-NOAA I added an `ATM_` prefix to the `BEGIN_YYYYmmddHHMMSS` in `yamltools.py` and `parm/atm/hofx/hofx4d.yaml`. Thanks.

Comment (jiaruidong2017) [reply ush/ufsda/yamltools.py:189]: > For the timing, the (possible) issue is that the land DA update is applied to a restart. So we should make the analysis time be the same as the restart time - which means it may not be the same as the atmos analysis time. For example, for IAU the restart is at the start of the window and I think (?) the (nominal) atmos analysis time is the middle of the window. In this case the land analysis time should be the start of the window (or, the end of the previous window).

@ClaraDraper-NOAA This is a key topic for adding our land analysis. We can continue this discussion before our next PR on adding the `add_increment` for land analysis into the GDASApp.

**Review by ClaraDraper-NOAA — APPROVED** [review]

**Review by CoryMartin-NOAA — CHANGES_REQUESTED** [review]
One thing remaining @jiaruidong2017 , since you changed it to `ATM_BEGIN_YYYYmmddHHMMSS`, this needs to be reflected everywhere in the parm/ directory (so all the atm files need updated or they will break with the change to the python script you made). Otherwise I think this is ready after that.

**Comment by jiaruidong2017** [comment]
> One thing remaining @jiaruidong2017 , since you changed it to `ATM_BEGIN_YYYYmmddHHMMSS`, this needs to be reflected everywhere in the parm/ directory (so all the atm files need updated or they will break with the change to the python script you made). Otherwise I think this is ready after that.

@CoryMartin-NOAA I found the only yaml file including the `BEGIN_YYYYmmddHHMMSS` is `parm/atm/hofx/hofx4d.yaml`, and I made the change from `BEGIN_YYYYmmddHHMMSS` to `ATM_BEGIN_YYYYmmddHHMMSS` and already committed the changes to the GDADSApp repo.

**Comment by CoryMartin-NOAA** [comment]
@jiaruidong2017 yes my mistake, I was conflating that with the BKG variable.

**Review by CoryMartin-NOAA — APPROVED** [review]
Thanks @jiaruidong2017 this is a good first step. This will be an iterative process and things will likely change but this is great so far!

**Pull request merged**
